lib/spreadsheet_exporter/xlsx.rb

```diff
@@ -1,43 +1,139 @@
-        values.each_with_index do |value, col|
-          worksheet.write(row + 1, col, value)
-        end
+      Array(spreadsheet[1..]).each_with_index do |values, row|
+        worksheet.write_row(row + 1, 0, Array(values))
+      end
+
+      add_worksheet_validation(workbook, worksheet, column_indexes, header_format, options)
+
+      workbook.worksheets.each do |ws|
+        ws.freeze_panes(1, 0)
       end
 
-      # Output the file contents and delete it
       workbook.close
-      file = File.open(temp_file_path)
-      output = file.read
-      file.close
-      File.delete(temp_file_path)
+      io.string
+    end
+
+    # TODO: we should DRY this up with the Spreadsheet.from_objects logic
+    def self.rewrite_validation_column_names(column_validations, options)
+      return column_validations unless options["humanize_headers_class"]
+      klass = options["humanize_headers_class"]
+
+      column_validations.each_with_object({}) do |(attribute, v), obj|
+        rewritten = klass.human_attribute_name(attribute)
+        rewritten.downcase! if options[:downcase]
+        obj[rewritten] = v
+      end
+    end
+
+    def self.add_worksheet_validation(workbook, worksheet, column_indexes, header_format, options = {})
+      column_validations = options.fetch("validations", {}) || {}
+      return if column_validations.empty?
+
+      column_validations = rewrite_validation_column_names(column_validations, options)
+
+      column_validations.each do |column_name, column_validation|
+        column_index = column_indexes[column_name]
+
+        if column_index.nil?
+          # TODO: we should output an empty column anyways
+          warn "attempted to apply validation to missing column '#{column_name}'"
+          next
+        end
+
+        validation_options = add_column_validation(workbook, column_name, column_index, column_validation, header_format)
+
+        pp validation_options
+
+        worksheet.data_validation(1, column_index, ROW_MAX, column_index, validation_options)
+      end
+    end
+
+    def self.add_column_validation(workbook, column_name, column_index, column_validation, header_format)
+      list_values = Array(column_validation.fetch("source", []))
+      if list_values.empty?
+        raise ArgumentError, "no values for validation for column '#{column_name}'"
+      end
+
+      error_type = column_validation.fetch("error_type", VALIDATION_ERROR_TYPES[0])
+      unless VALIDATION_ERROR_TYPES.include?(error_type)
+        raise ArgumentError, "invalid error_type `#{error_type}` for validation for column '#{column_name}'"
+      end
+
+      list_values.compact!
+      list_length = list_values.join(",").length
+
+      source = nil
+
+      if USE_INLINE_LISTS && list_length <= MAX_INLINE_LIST_CHARS
+        # commas are not allowed when
+        # TODO: we should warn about losing any commas
+        list_values.map! { |v| v.sub(',', '').strip }
+        source = list_values
+      else
```

Comment:
Write the list of values to the `DATA_WORKSHEET_NAME` worksheet instead of inline